### PR TITLE
Typecheck nested properties

### LINKED ISSUE #330: Typechecker allows non-existent fields in embedded objects
### What happened?

See https://github.com/pulumi/pulumi/issues/10466 for details.

### Steps to reproduce

Run `pulumi up` on 
```yaml
name: yaml-test
runtime: yaml
description: A minimal Pulumi YAML program
resources:
  patch1:
    type: kubernetes:apps/v1:DeploymentPatch
    properties:
      metadata:
        name: "coredns"
        namespace: "kube-system"
        annotations:
          pulumi.com/patchForce: "true"
      spec:
        template:
          containers:
            name: "coredns"
            # Use the version retrieved earlier as the image tag here!
            image: "602401143452.dkr.ecr.us-west-2.amazonaws.com/eks/coredns:v1.8.7-eksbuild.1"
```
With `pulumi-eks = v0.41.2`

### Expected Behavior

We should reject an invalid program.

### Actual Behavior

We let it in.

### Output of `pulumi about`

_No response_

### Additional context

_No response_

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

### PR DESCRIPTION
**Pull request by aq17**
Fixes #330 

todo: add test

### CODE REVIEW

File: pkg/pulumiyaml/analyser.go [hunk 1]
```diff
@@ -490,6 +492,29 @@ func (tc *typeCache) typePropertyEntries(ctx *evalContext, resourceName string,
 			subject := kvp.Key.Syntax().Syntax().Range()
 			ctx.addErrDiag(subject, summary, detail)
 		} else {
+			if obj, ok := kvp.Value.(*ast.ObjectExpr); ok {
+				prop := propMap[kvp.Key.Value]
```
Comment (iwahbe) [inline pkg/pulumiyaml/analyser.go:496]: This variable already exists as `typ`. It's bound during the if check.

**Review by iwahbe — CHANGES_REQUESTED** [review]
I don't think that `typePropertyEntries` is the right place to put this. It serves to aggregate properties into a `schema.Type` and hand that off to `assertTypeAssignable`. I think the fix needs to live in `assertTypeAssignable`.

I'm pretty sure there is a bug in 

https://github.com/pulumi/pulumi-yaml/blob/60bfab81632daf5cb945e04b57412dc135fed4d8/pkg/pulumiyaml/analyser.go#L295-L321

`assertTypeAssignable` handles complications like `schema.Union`, which we need to account for.

**Comment by aq17** [comment]
> I don't think that `typePropertyEntries` is the right place to put this. It serves to aggregate properties into a `schema.Type` and hand that off to `assertTypeAssignable`. I think the fix needs to live in `isAssignable`.
> 
> I'm pretty sure there is a bug in
> 
> https://github.com/pulumi/pulumi-yaml/blob/60bfab81632daf5cb945e04b57412dc135fed4d8/pkg/pulumiyaml/analyser.go#L295-L321
> 
> `isAssignable` handles complications like `schema.Union`, which we need to account for.

I'm unsure how you get the subproperties (`from`) within just `isAssignable` though. If the`from` property is an object, we can get its subproperties and recursively call `isAssignable` (where `from` is now `to`), but what do we compare it to?

**Comment by iwahbe** [comment]
`to` would also be an object, `isAssignable` just asserts that an object fits within another object.

**Pull request closed without merging**